src/main/java/com/wojciechkocik/usage/service/TimeSpentCrossMidnightServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+package com.wojciechkocik.usage.service;
+
+import com.wojciechkocik.usage.dto.DailyUsage;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalTime;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of {@link TimeSpentCrossMidnightService} using java 8 features
+ *
+ * @author Wojciech Kocik
+ * @since 24.03.2017
+ */
+@Service
+public class TimeSpentCrossMidnightServiceImpl implements TimeSpentCrossMidnightService {
+
+    @Override
+    public List<DailyUsage> run(List<DailyUsage> dailyUsagesForCourse) {
+
+        List<DailyUsage> newDaysForMergeWithCurrentList = new ArrayList<>();
+
+        dailyUsagesForCourse.forEach(p -> {
+            ZonedDateTime dateTime = p.getDateTime();
+
+            int dayBeforeSessionDuration = dateTime.getDayOfMonth();
+
+            long sessionDuration = p.getTime();
+            int dayAfterSessionDuration = dateTime.plusSeconds(sessionDuration).getDayOfMonth();
```

Comment:
Here is a subtle bug. It is very improbable, but in some cases you will get wrong results (when the sessionDuration took around 30 days).